### PR TITLE
Refinements to input-source specific table functions

### PR DESCRIPTION
**Pull request by paul-rogers**
This is a follow-up to the earlier PR that introduced input-source specific table functions (e.g. `http`, `s3`, `localfiles`, etc.) This PR is mostly cleanup and bug fixes and is a spin-off of the effort to integrate the catalog with Calcite. These changes are done as a separate PR to keep each PR to a reasonable size.

* Fixes various bugs
* Improves the structure of the table function classes 
* Adds unit and integration tests

In particular, this PR fixes an issue with the feature that allows an SQL feature with the `extern` function.

#### Release note

No user-visible changes: just refinements to functionality already described in an earlier PR.

<hr>

This PR has:

- [X] been self-reviewed.
- [X] a release note entry in the PR description.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [X] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java [hunk 1]
```diff
@@ -238,6 +241,7 @@ protected InputSource convertSource(
       return jsonMapper.convertValue(jsonMap, inputSourceClass());
     }
     catch (Exception e) {
+      LOG.debug(e, "Invalid input source specification");
```
Comment (zachjsh) [inline server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java:244]: Should this be error log? debug isnt logged by default.

File: server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java [hunk 2]
```diff
@@ -95,7 +95,10 @@ protected List<ParameterDefn> addFormatParameters(
   )
   {
     final List<ParameterDefn> toAdd = new ArrayList<>();
-    final ParameterDefn formatProp = new Parameter(FORMAT_PARAMETER, ParameterType.VARCHAR, false);
+    // While the format parameter is required, we mark it as optional. Else
+    // if the source defines optional parameters, they will still be ignored
+    // as Calcite treats (optional, optional, required) as (required, required, required)
```
Comment (zachjsh) [inline server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java:100]: What do you mean by this? Do you mean that if any of the parameters are required then all are?

File: server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java [hunk 3]
```diff
@@ -95,7 +95,10 @@ protected List<ParameterDefn> addFormatParameters(
   )
   {
     final List<ParameterDefn> toAdd = new ArrayList<>();
-    final ParameterDefn formatProp = new Parameter(FORMAT_PARAMETER, ParameterType.VARCHAR, false);
+    // While the format parameter is required, we mark it as optional. Else
+    // if the source defines optional parameters, they will still be ignored
```
Comment (zachjsh) [inline server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java:99]: if the format* defines optional parameters?

File: server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java (hunk 1 above)
Comment (paul-rogers) [reply server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java:244]: This is for debugging. Specifically, the case in which an input format requires a JSON or Guice thing which is not configured in the Coordinator. Only going to happen for developers, not for users.

File: server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java (hunk 3 above)
Comment (paul-rogers) [reply server/src/main/java/org/apache/druid/catalog/model/table/FormattedInputSourceDefn.java:99]: Actually, _input source_, as written. The input source parameters preceded the input format parameters. That's how we end up with the scenario in the comment: the input source has two optional parameters, but the format has a required one. That forces the input source parameters to also become required.

This is, by the way, why that bug in `extern`, where I had the optional flag backward, actually worked. Calcite took my `(optional, optional, required)` and turned it into what the code originally had: `(required, required, required)`.

**Review by zachjsh — APPROVED** [review]
LGTM 🚀

**Pull request merged**